cfgov/login/auth.py

```diff
@@ -0,0 +1,137 @@
+    except Group.DoesNotExist:
+        logger.debug(
+            "Default group 'Wagtail Users' does not exist, "
+            "not modifying user groups"
+        )
+        return
+
+    if not user.groups.contains(group):
+        logger.debug(f"Adding {user.username} to {group.name}.")
+        user.groups.add(group)
+
+
+def process_roles_admin(user, roles):
+    """Adjust a user's is_superuser property based on OIDC roles claim"""
+    if settings.OIDC_OP_ADMIN_ROLE is None:
+        logger.debug("OIDC_OP_ADMIN_ROLE not set, not modifying is_superuser")
+        return
+
+    is_admin = settings.OIDC_OP_ADMIN_ROLE in roles
+
+    if is_admin and not user.is_superuser:
+        logger.debug(f"Setting is_superuser for {user.username}.")
+        user.is_superuser = True
+        return True
+    elif not is_admin and user.is_superuser:
+        logger.debug(f"Removing is_superuser from {user.username}.")
+        user.is_superuser = False
+        return True
+
+
+def process_roles_claim(user, roles):
+    """Adjust is_superuser and user group assignment based on OIDC roles"""
+    # All SSO users get access to the Wagtail admin.
+    # This function does not modify user objects directly, only group
+    # membership.
+    ensure_wagtail_users_group_membership(user)
+
+    # Specific users should have superuser status based on the contents of
+    # their roles
+    user_modified = process_roles_admin(user, roles)
+    return bool(user_modified)
+
+
+def process_given_name_claim(user, given_name):
+    """Set a Django user's first_name to an OIDC given_name"""
+    if user.first_name != given_name:
+        user.first_name = given_name
+        return True
+
+
+def process_family_name_claim(user, family_name):
+    """Set a Django user's last_name to an OIDC family_name"""
+    if user.last_name != family_name:
+        user.last_name = family_name
+        return True
+
+
+class CFPBOIDCAuthenticationBackend(OIDCAuthenticationBackend):
+
+    def get_userinfo(self, access_token, id_token, payload):
+        # Roles are part of the parent payload here, and not the userinfo.
+        # Preserve them if we can.
+        roles = payload.get("roles", [])
+        userinfo = super().get_userinfo(access_token, id_token, payload)
+        userinfo["roles"] = roles
+        return userinfo
+
+    def process_claims(self, user, claims):
+        """Run a configured callable for each OIDC claim received
+
+        If a claim processor modifies a user it should return True, and the
+        user will be saved once all claims are processed.
+        """
+        user_modified = False
+
+        claim_processors = getattr(settings, "OIDC_CLAIMS_PROCESSORS", {})
+        for claim, processor_string in claim_processors.items():
+            if claim not in claims:
+                continue
+
+            processor = import_string(processor_string)
+            user_modified |= bool(processor(user, claims[claim]))
```

Comment:
This may be a nitpicky suggestion, but re-reading this, I wonder if adding a new setting for OIDC_CLAIMS_PROCESSORS is worth it. Originally I thought that was a convention for the Mozilla package but I don't see it there. With this implementation we're adding yet another new OIDC setting -- we're already defining many new settings there -- but it's not one that will be varying by environment or changing often.

This file is importing that setting which references method names in this same file, which we then reimport, even though we're already in this file. Is it worth the simplification to instead do something like


```suggestion
        claim_processors = {
            "roles": process_roles_claim,
            "given_name": process_given_name_claim,
            "family_name": process_family_name_claim,
        }
        
        for claim_type, claim in claims.items():
            if processor := claim_processors.get(claim_type):
                user_modified |= bool(processor(user, claim))
```

Reply:
Maybe... this would simplify settings, and still enable us to do something more robust there if we want to in the future. 

The individual claim functions are separate themselves so they're nice, small, independently testable. I thought it might be nice to make it configurable. But if the settings are overwhelming, we can drop that and do as you suggest. I'm well in the weeds of OIDC (another library does it this way, in configuration mapping to code), so that's where it's valuable to know if that's too much.

Reply:
Done in [986b7ea](https://github.com/cfpb/consumerfinance.gov/pull/8414/commits/986b7ea7114be6f71943d3f4517634a004158b04)